cmd/kubeadm/app/cmd/init.go

```diff
@@ -397,14 +392,20 @@ func (i *Init) Run(out io.Writer) error {
 
 	// PHASE 6: Install and deploy all addons, and configure things as necessary
 
-	if err := dnsaddonphase.EnsureDNSAddon(i.cfg, client); err != nil {
-		return err
-	}
-
 	if err := proxyaddonphase.EnsureProxyAddon(i.cfg, client); err != nil {
 		return err
 	}
 
+	// Make the service discovery CoreDNS if feature gate is enabled
+	if features.Enabled(i.cfg.FeatureGates, features.CoreDNS) {
+		if err := dnsaddonphase.EnsureCoreDNSAddon(i.cfg, client); err != nil {
+			return err
+		}
+	} else {
+		if err := dnsaddonphase.EnsureDNSAddon(i.cfg, client); err != nil {
```

Comment:
Can you move this logic inside of `EnsureDNSAddon`? (i.e. the logic whether to use kube- or core-dns)